File: adaptive/src/main/java/com/example/adaptive/MainActivity.java
package com.example.adaptive;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import com.bumptech.glide.Glide;
import com.example.adaptive.databinding.ActivityMainBinding;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityMainBinding binding;
        binding = ActivityMainBinding.inflate(getLayoutInflater());

        setContentView(binding.getRoot());

        binding.btn0.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkPermission();
            }
        });

        binding.btn1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                //这里只是读取，但是在内部存储的目录下，还是创建了一个files文件夹（但是文件myFiles依旧没创建）
                String path = StoreUtils.getFilePath(MainActivity.this);
                //路径：/data/user/0/com.example.adaptive/files/myFile
                Log.d("Daisy", "获取到的路径:" + path);

                StoreUtils.writeFile(path);
            }
        });

        binding.btn2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String path = StoreUtils.getFilePath(MainActivity.this);

                StoreUtils.readFile(path);
            }
        });

        binding.btn3.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.usePackagingOpenFileOutput(MainActivity.this);
            }
        });

        binding.btn4.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.usePackingOpenFileInput(MainActivity.this);
            }
        });

        binding.btn5.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.testSp("sp_test", "hsf", "黄绍飞", MainActivity.this);
            }
        });

        binding.btn6.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.testDb(MainActivity.this);
            }
        });

        binding.btn7.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.testGetCodeCacheDir(MainActivity.this);
            }
        });

        binding.btn8.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                StoreUtils.testGetDataDir(MainActivity.this);
            }
        });



        binding.btn9.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String lastPic = ExternalUtils.getImagePath(MainActivity.this);
//                String lastPic = "/storage/emulated/0/Pictures/Screenshots/Screenshot_2023-04-17-18-23-34-18_651b5c74a671b75cef00c5b91170e3c1.jpg";

                //用Glide可以展示外部存储的图片
                /* 如果没有外部存储权限，则会报以下错
                 open failed: EACCES (Permission denied)
                 */
                Glide.with(MainActivity.this).load(lastPic).into(binding.ivShow);
            }
        });

    }


    /**
     * 申请外部存储权限
     */
    public void checkPermission() {
        List<String> permissionsToRequire = new ArrayList<>();

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            permissionsToRequire.add(Manifest.permission.READ_EXTERNAL_STORAGE);
        }

        String[] ppp  = permissionsToRequire.toArray(new String[permissionsToRequire.size()]);
        if (!permissionsToRequire.isEmpty()) {
            ActivityCompat.requestPermissions(this, ppp, 0);
        }
    }
}

File: app/src/main/java/com/example/storagetest/MainActivity.kt
package com.example.storagetest

import android.Manifest
import android.content.ContentUris
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.provider.MediaStore.Audio.Media
import android.util.Log
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.example.storagetest.databinding.ActivityMainBinding
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.net.HttpURLConnection
import java.net.URL
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        checkPermission()

        binding.btnTest.setOnClickListener {
            val cursor = contentResolver.query(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, null, null, null, "${MediaStore.MediaColumns.DATE_ADDED} desc")
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    val id = cursor.getLong(cursor.getColumnIndexOrThrow(MediaStore.MediaColumns._ID))
                    val uri = ContentUris.withAppendedId(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, id)
                    Log.d("Daisy", "image uri is $uri")
                }
                cursor.close()
            }
        }

        binding.btnBrowseAlbum.setOnClickListener {
            val intent = Intent(this, BrowseAlbumActivity::class.java)
            startActivity(intent)
        }

        binding.btnAddPhoto.setOnClickListener {
            val bitmap = BitmapFactory.decodeResource(resources, R.drawable.image)
            val displayName = "hsf.jpg"
            val mineType ="image/jpeg"
            val compressFormat = Bitmap.CompressFormat.JPEG

            val values = ContentValues()
            values.put(MediaStore.MediaColumns.DISPLAY_NAME, displayName)
            values.put(MediaStore.MediaColumns.MIME_TYPE, mineType)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                values.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_DCIM)
            } else {
                //用的是路径拼接
                values.put(MediaStore.MediaColumns.DATA, "${Environment.getExternalStorageDirectory().path}/" +
                        "${Environment.DIRECTORY_DCIM}/${displayName}")
            }
            val uri = contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, values)
            if (uri != null) {
                val outputStream = contentResolver.openOutputStream(uri)
                if (outputStream != null) {
                    bitmap.compress(compressFormat, 100, outputStream)
                    outputStream.close()
                    Toast.makeText(this, "Add bitmap to album succeeded", Toast.LENGTH_SHORT).show()
                }
            }
        }

        binding.btnDownload.setOnClickListener {
            val fileUrl = "http://guolin.tech/android.txt"
            val fileName = "android_hsf.txt"
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
                Toast.makeText(this, "你必须使用大于或等于Android10的设备", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            thread {
                try {
                    val url = URL(fileUrl)
                    val connection = (url.openConnection() as HttpURLConnection).also {
                        it.requestMethod = "GET"
                        it.connectTimeout = 8000
                        it.readTimeout = 8000
                    }
                    val inputStream = connection.inputStream
                    val bis = BufferedInputStream(inputStream)
                    val values = ContentValues().apply {
                        put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
                        put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_DOWNLOADS)
                    }
                    val uri = contentResolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, values)
                    if (uri != null) {
                        val outputStream = contentResolver.openOutputStream(uri)
                        if (outputStream != null) {
                            val bos = BufferedOutputStream(outputStream)
                            val buffer = ByteArray(1024)
                            var bytes = bis.read(buffer)
                            while (bytes >= 0) {
                                bos.write(buffer, 0, bytes)
                                bos.flush()
                                bytes = bis.read(buffer)
                            }
                            bos.close()
                            runOnUiThread {
                                Toast.makeText(this, "下载完毕", Toast.LENGTH_SHORT).show()
                            }
                        }

                    }
                    bis.close()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

    }

    private fun checkPermission() {
        val permissionsToRequire = ArrayList<String>()
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            permissionsToRequire.add(Manifest.permission.READ_EXTERNAL_STORAGE)
        }
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            permissionsToRequire.add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
        }
        if (permissionsToRequire.isNotEmpty()) {
            ActivityCompat.requestPermissions(this, permissionsToRequire.toTypedArray(), 0)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 0) {
            for (result in grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    Toast.makeText(this, "You must allow all the permissions.", Toast.LENGTH_SHORT).show()
                    finish()
                }
            }
        }
    }
}